src/database/listRemote.ts

```diff
@@ -0,0 +1,78 @@
+import * as request from "request";
+import { Response } from "request";
+import * as responseToError from "../responseToError";
+import * as utils from "../utils";
+import * as FirebaseError from "../error";
+import * as logger from "../logger";
+import * as api from "../api";
+
+export interface ListRemote {
+  /**
+   * Call the shallow get API with limitToFirst=numSubPath.
+   * @param path the path to list
+   * @param numSubPath the number of subPaths to fetch.
+   * @return the list of sub pathes found.
+   */
+  listPath(path: string, numSubPath: number, offset?: string, timeout?: number): Promise<string[]>;
+}
+
+export class RTDBListRemote implements ListRemote {
+  private instance: string;
+
+  constructor(instance: string) {
+    this.instance = instance;
+  }
+
+  listPath(path: string, numSubPath: number, offset?: string, timeout?: number): Promise<string[]> {
+    const offsetSuffix = offset ? "&startAfter=" + offset : "";
+    const timeoutSuffix = timeout ? "&timeout=" + timeout + "ms" : "";
+    const url =
+      utils.addSubdomain(api.realtimeOrigin, this.instance) +
+      path +
+      `.json?shallow=true&limitToFirst=${numSubPath}` +
+      offsetSuffix +
+      timeoutSuffix;
+
+    const t0 = Date.now();
+    return api
+      .addRequestHeaders({
+        url,
+      })
+      .then((reqOptionsWithToken) => {
```

Comment:
this is typescript, so we can use async/await notation to simplify this a bit

Reply:
Let me know if the latest commit addresses your concern here.